apm-agent-core/src/main/java/co/elastic/apm/agent/bci/ElasticApmAgent.java

```diff
@@ -753,48 +755,64 @@ public static File getAgentJarFile() {
-                ElasticApmTracer tracer = GlobalTracer.requireTracerImpl();
+                final ElasticApmTracer tracer = GlobalTracer.requireTracerImpl();
                 if (instrumentation == null) {
                     throw new IllegalStateException("Agent is not initialized");
                 }
 
-                appliedInstrumentations = dynamicallyInstrumentedClasses.get(classToInstrument);
-                if (!appliedInstrumentations.contains(instrumentationClasses)) {
-                    appliedInstrumentations = new HashSet<>(appliedInstrumentations);
-                    appliedInstrumentations.add(instrumentationClasses);
-                    // immutability guards against race conditions (for example concurrent rehash due to add and lookup)
-                    appliedInstrumentations = Collections.unmodifiableSet(appliedInstrumentations);
-                    dynamicallyInstrumentedClasses.put(classToInstrument, appliedInstrumentations);
-
-                    CoreConfiguration config = tracer.getConfig(CoreConfiguration.class);
-                    final Logger logger = LoggerFactory.getLogger(ElasticApmAgent.class);
-                    final ByteBuddy byteBuddy = new ByteBuddy()
-                        .with(TypeValidation.of(logger.isDebugEnabled()))
-                        .with(FailSafeDeclaredMethodsCompiler.INSTANCE);
-                    AgentBuilder agentBuilder = getAgentBuilder(
-                        byteBuddy, config, logger, AgentBuilder.DescriptionStrategy.Default.POOL_ONLY, false, false
-                    );
-                    for (Class<? extends ElasticApmInstrumentation> instrumentationClass : instrumentationClasses) {
-                        ElasticApmInstrumentation apmInstrumentation = instantiate(instrumentationClass);
-                        mapInstrumentationCL2adviceClassName(
-                            apmInstrumentation.getAdviceClassName(),
-                            PrivilegedActionUtils.getClassLoader(instrumentationClass));
-                        ElementMatcher.Junction<? super TypeDescription> typeMatcher = getTypeMatcher(classToInstrument, apmInstrumentation.getMethodMatcher(), none());
-                        if (typeMatcher != null && isIncluded(apmInstrumentation, config)) {
-                            agentBuilder = applyAdvice(tracer, agentBuilder, apmInstrumentation, typeMatcher.and(apmInstrumentation.getTypeMatcher()));
-                        }
+                final Set<Collection<Class<? extends ElasticApmInstrumentation>>> updatedAppliedInstrumentations =
+                    dynamicallyInstrumentedClasses.get(classToInstrument);
+                if (!updatedAppliedInstrumentations.contains(instrumentationClasses)) {
+                    if (System.getSecurityManager() == null) {
+                        applyInstrumentation(classToInstrument, instrumentationClasses, updatedAppliedInstrumentations, tracer);
+                    } else {
```

Comment:
[minor] do you remember what method(s) triggered security manager checks here ? Maybe a small comment might help to understand why it's required here (and could also help answer questions we may have about all the "privileged code" we have).

Reply:
I'll look for it

Reply:
The first one to fail is:
```
[tomcat] java.security.AccessControlException: access denied ("java.io.FilePermission" "/elastic-apm-agent.jar" "read")
[tomcat] at java.security.AccessControlContext.checkPermission(AccessControlContext.java:472)
[tomcat] at java.security.AccessController.checkPermission(AccessController.java:684)
[tomcat] at java.lang.SecurityManager.checkPermission(SecurityManager.java:549)
[tomcat] at java.lang.SecurityManager.checkRead(SecurityManager.java:888)
[tomcat] at java.util.zip.ZipFile.<init>(ZipFile.java:212)
[tomcat] at java.util.zip.ZipFile.<init>(ZipFile.java:151)
[tomcat] at java.util.jar.JarFile.<init>(JarFile.java:154)
[tomcat] at java.util.jar.JarFile.<init>(JarFile.java:118)
[tomcat] at net.bytebuddy.dynamic.ClassFileLocator$ForJarFile.of(ClassFileLocator.java:782)
[tomcat] at co.elastic.apm.agent.bci.ElasticApmAgent.getAgentBuilder(ElasticApmAgent.java:665)
[tomcat] at co.elastic.apm.agent.bci.ElasticApmAgent.applyInstrumentation(ElasticApmAgent.java:791)
[tomcat] at co.elastic.apm.agent.bci.ElasticApmAgent.ensureInstrumented(ElasticApmAgent.java:772)
...
```
I assume if I add a specific permission for that I would find the second, however I think it is enough to add a general comment, which I will.